Load device profile on import

nicholaspcr opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

References #5512 
Blocked by #5659

The main idea of this is to apply the end profile loading on the back-end, that way there can be some sort of control on the front-end (controlled through the injection of the `application_id` to the request). Therefore this doesn't close the issue but adds the functionality. There still needs to be the UI change on the front-end but the CLI should work.

#### Changes
<!-- What are the changes made in this pull request? -->

- Add `vendor_id` and and `vendor_profile_id` fields to `csv` import
- Add `vendor_id` and `vendor_profile_id` to allowed field mask list on end_device creation.
- Add `EndDeviceProfileIdentifiers` to convert service. Allows for setting a fallback value for fetching templates for all devices added.
- Add more test cases for `ttscsv`
- Add profile fetcher to DTC and unit tests for it.

#### Testing

<!-- How did you verify that this change works? -->

Unit tests and manual testing.
Manual testing:
- Temporarily add the `application_id` field to the converter and use the console.
- Run the template through the CLI.

##### Regressions

<!-- Please indicate features that this change could affect and how that was tested. -->

The old way of handling the request should still work, this is added in a way that the loading only happen if the user specify the `vendor_id` and `vendor_profile_id` on the CSV file or if the console provide the fallback values in the request.

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

Will add the documentation after this is approved.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

nicholaspcr committed:
util,dev: Add vendor fields to csv converter setters

nicholaspcr committed:
dev: Move DR mock to its own package

adriansmares commented on pkg/devicetemplateconverter/devicetemplateconverter.go:
```diff
@@ -45,12 +46,28 @@ var errNotFound = errors.DefineNotFound("converter", "converter `{id}` not found
+		devicetemplates.RegisterConverter(
+			devicetemplates.TTSCSV,
+			devicetemplates.NewCSV(&ProfileFetcher{
+				DeviceRepositoryAddress: conf.DeviceRepositoryAddress,
```
You should use the peering functionality (allow `ProfileFetcher` to use `GetPeerConn`) instead. Carrying the address internally has its own issues (like missing `InsecureSkipVerify`).

adriansmares commented on pkg/devicetemplateconverter/profile_fetcher.go:
```diff
@@ -0,0 +1,56 @@
+
+// FetchProfile provides the end device profile.
+func (pf *ProfileFetcher) FetchProfile(ctx context.Context, device *ttnpb.EndDevice) (*ttnpb.EndDeviceTemplate, error) {
+	conn, err := grpc.Dial(pf.DeviceRepositoryAddress, grpc.WithInsecure())
```
Also because you're manually calling, you don't have the `rpcclient` options here.

nicholaspcr committed:
dev: Add vendorID fields to allowed paths validation on device creation

johanstokking commented on api/end_device.proto:
```diff
@@ -860,4 +860,7 @@ message ConvertEndDeviceTemplateRequest {
   // Data to convert.
   bytes data = 2;
+  
+  // Mandatory when loading the device profile from the Device Repository.
+  ApplicationIdentifiers application_ids = 3;
```
Yeah so let's get rid of this, see https://github.com/TheThingsNetwork/lorawan-stack/issues/5512#issuecomment-1189511262

nicholaspcr replied:
@johanstokking was working on deprecating the appID on DR services and updating the auth validation in this PR but it easily became its own thing, meaning that it impacts quite a few tests. So I ended up making this PR a draft and  will create a separate PR with those changes. That should make it easier to review.

johanstokking reviewed (commented):
Very nice. Minor comment, also there are a couple of valid lint warnings and errors.

nicholaspcr committed:
dev,util: Add fetcher tests to csv converter

nicholaspcr committed:
dev: Update CHANGELOG

adriansmares commented:
Since you're chaining PRs, you can make this one target `feature/remove-app-id-from-dr` in order to make things easier to review.

johanstokking commented on api/identifiers.proto:
```diff
@@ -150,6 +150,16 @@ message EndDeviceVersionIdentifiers {
+  // ID of the LoRaWAN end device profile assigned by the vendor.
+  uint32 vendor_profile_id = 2;
+}
+
+
```
Remove

johanstokking commented on pkg/devicetemplateconverter/devicetemplateconverter.go:
```diff
@@ -88,5 +105,7 @@ func (dtc *DeviceTemplateConverter) RegisterServices(s *grpc.Server) {
 // RegisterHandlers registers gRPC handlers.
 func (dtc *DeviceTemplateConverter) RegisterHandlers(s *runtime.ServeMux, conn *grpc.ClientConn) {
-	ttnpb.RegisterEndDeviceTemplateConverterHandler(dtc.Context(), s, conn)
+	if err := ttnpb.RegisterEndDeviceTemplateConverterHandler(dtc.Context(), s, conn); err != nil {
+		panic(err)
+	}
```
Why?

nicholaspcr replied:
I was running the golintci locally to catch errors that would show up on the CI, one of them was not checking the errors of methods called in the `RegisterHandlers`. The panic was due to this being called at the server initialization, so it made sense to me stopping the server if the handlers aren't weren't able to be registered.

Nevertheless the other places indeed don't do this so might not be appropriate to do so here, will remove it.

johanstokking commented on pkg/devicetemplateconverter/profile_fetcher.go:
```diff
@@ -0,0 +1,57 @@
+	CallOpts    func(context.Context) ([]grpc.CallOption, error)
+}
+
+// ValidateFields dictactes if the end device has the necessary fields to fetch its profile.
+func (pf *ProfileFetcher) ValidateFields(device *ttnpb.EndDevice) bool {
```
This method name indicates it's about validation and about validation by fields. This is not what this method does, right? This just indicates whether the profile is set?

nicholaspcr replied:
Effectively there is validation, although simple,  of end-device's fields that are necessary for fetching its profile from the DR. I assume it would be more intuitive for it to be named something like `IsFetchable`.

adriansmares replied:
`IsFetchable` gives the impression that the fetcher (the receiver) is 'fetchable'. `ShouldFetchProfile` is most likely closer to what you want.

johanstokking commented on pkg/devicetemplateconverter/profile_fetcher.go:
```diff
@@ -0,0 +1,57 @@
+		log.FromContext(ctx).WithError(err).Warn("Failed to get Device Repository peer")
+		return nil, err
+	}
+	opts, err := pf.CallOpts(ctx)
```
Wouldn't we naturally forward the auth here anyway?

nicholaspcr replied:
It is called but not so by conventional means, [here](https://github.com/TheThingsNetwork/lorawan-stack/blob/cc4e6e5928cb0adbb4c6a8959c915ea88265dd6b/pkg/devicetemplateconverter/devicetemplateconverter.go#L55-L61) its where the callOpts for the profile fetcher are specified. I tried to make something a bit more customization in order to facilitate testing of the fetcher. This was something that I should've mentioned in the notes for reviewers, since it is not the standard way of doing so. If you feel it is inappropriate I'll revert it.

KrishnaIyer replied:
The way we do this is by defining the target client. See https://github.com/TheThingsNetwork/lorawan-stack/blob/cb215d00b2f1563f9620c653a0ed223025de71ff/pkg/deviceclaimingserver/enddevices/enddevices.go#L254-L260

I don't think adding the `callopts` to the constructor is the correct way to go. With the suggested method, the mock just needs to implement the client interface functions instead of needing to be an RPC server.

adriansmares commented on api/identifiers.proto:
```diff
@@ -150,6 +150,16 @@ message EndDeviceVersionIdentifiers {
+  // ID of the LoRaWAN end device profile assigned by the vendor.
+  uint32 vendor_profile_id = 2;
+}
+
```
Only one new line.

adriansmares commented on api/devicerepository.proto:
```diff
@@ -298,16 +298,7 @@ message ListEndDeviceModelsRequest {
-  }
-
-  // Application identifiers.
+    // Application identifiers.
```
Remove the two extra spaces.

adriansmares commented on cmd/ttn-lw-cli/commands/end_device_templates.go:
```diff
@@ -37,6 +37,13 @@ func templateFormatIDFlags() *pflag.FlagSet {
 	return flagSet
 }
 
+func endDeviceProfileIDFlags() *pflag.FlagSet {
```
This is inconsistent with the field naming in the request message. The field is called `end-device-profile-ids`, not `vendor-ids`.

You don't need to manually write this flag set creator. If you add
```
  option (thethings.flags.message) = { set: true };
```
to the `EndDeviceProfileIdentifiers` and `ConvertEndDeviceTemplateRequest` messages, both this and the `templateFormatIDFlags` will be generated for you. See the resulting `AddSetFlagsForConvertEndDeviceTemplateRequest` function `pkg/ttnpb`.

Also consider using `ConvertEndDeviceTemplateRequest.SetFromFlags` instead of `getEndDeviceProfileID`.

nicholaspcr replied:
I ended only applying this to the `EndDeviceProfileIdentifiers`, the reason for this is that when loading the values of `ConvertEndDeviceTemplateRequest` the slice of bytes assigned to the `data` field is expected but that is fetched through a `--local-file` flag which read the local file. [code here](https://github.com/TheThingsNetwork/lorawan-stack/blob/5b37ddd1e6bc9a86f07bf8a43748ad96500968e2/cmd/ttn-lw-cli/commands/flags.go#L237-L250)

adriansmares commented on pkg/devicetemplateconverter/config.go:
```diff
@@ -17,5 +17,6 @@ package devicetemplateconverter
 type Config struct {
-	Enabled []string `name:"enabled" description:"Enabled converters"`
+	Enabled                 []string `name:"enabled" description:"Enabled converters"`
+	DeviceRepositoryAddress string   `name:"device_repository_address" description:"Address of the Device Repository server"`
```
This not needed. You may safely assume there is always a DR in the cluster.

adriansmares commented on pkg/devicetemplateconverter/profile_fetcher.go:
```diff
@@ -0,0 +1,57 @@
+		return nil, err
+	}
+
+	return ttnpb.NewDeviceRepositoryClient(conn).GetTemplate(context.Background(), &ttnpb.GetTemplateRequest{
```
Why are you using the background context, and why are the application identifiers still provided given that they are deprecated ?

nicholaspcr replied:
Was me being careless, removed it

adriansmares commented on pkg/devicetemplateconverter/profile_fetcher_test.go:
```diff
@@ -0,0 +1,168 @@
+	"google.golang.org/grpc"
+)
+
+func Test_profileFetcher_ValidateFields(t *testing.T) {
```
`ProfileFetcher`, also below.

adriansmares commented on pkg/devicetemplateconverter/profile_fetcher_test.go:
```diff
@@ -0,0 +1,168 @@
+		validateErr  func(error) bool
+	}{
+		{
+			name: "fail/no_vendor_id",
```
Just FYI - `t.Run` will automatically convert spaces to underlines, so you may just use spaces if you wish here.

nicholaspcr replied:
I know but I usually have the preference of explicitly typing the underscore. Happened automatically, will make it similar to the other tests naming.

adriansmares commented on cmd/ttn-lw-cli/commands/end_device_templates.go:
```diff
@@ -295,6 +307,7 @@ This command takes end device templates from stdin.`,
 		Short:             "Convert data to an end device template (EXPERIMENTAL)",
 		PersistentPreRunE: preRun(),
 		RunE: func(cmd *cobra.Command, args []string) error {
+			optionalAuth()
```
Use `_ = func()` to discard the error explicitly.

adriansmares commented on cmd/ttn-lw-cli/commands/end_device_templates.go:
```diff
@@ -308,10 +321,15 @@ This command takes end device templates from stdin.`,
+				EndDeviceProfileIds: &ttnpb.EndDeviceProfileIdentifiers{},
+			}
+			// The execution should not stop if the flags are not found, therefore the error is ignored.
+			req.EndDeviceProfileIds.SetFromFlags(cmd.Flags(), "version-ids")
```
Same here.

adriansmares commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -33,7 +33,23 @@ import (
+	profile EndDeviceProfileFetcher
+}
+
+// ProfileCheck validates if its possible to fetch the end device profile from DR.
```
Fix comment.

adriansmares commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -33,7 +33,23 @@ import (
+
+// NewCSV allows the overwrite of the csv converter with an option interface that validates if it should fetch the End
+// Device profile from the Device Repository.
+func NewCSV(pc EndDeviceProfileFetcher) *ttsCSV {
```
Return `Converter`.

adriansmares commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -393,6 +443,19 @@ func (*ttsCSV) Convert(ctx context.Context, r io.Reader, ch chan<- *ttnpb.EndDev
+			if err := profileTmpl.EndDevice.SetFields(tmpl.EndDevice, tmpl.FieldMask.Paths...); err != nil {
+				return err
+			}
+			profileTmpl.FieldMask.Paths = append(profileTmpl.FieldMask.Paths, tmpl.FieldMask.Paths...)
```
Use `ttnpb.AddFields`.

adriansmares reviewed (approved):
🎉

KrishnaIyer commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -33,7 +33,24 @@ import (
+	profile EndDeviceProfileFetcher
+}
+
+// EndDeviceProfileFetcher specifies the necessary methods to ascertains if the end device contains the necessary
```
`ascertain`

KrishnaIyer commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -33,7 +33,24 @@ import (
+	FetchProfile(context.Context, *ttnpb.EndDevice) (*ttnpb.EndDeviceTemplate, error)
+}
+
+// NewCSV allows the overwrite of the csv converter with an option interface that validates if it should fetch the End
```
I think this can be simpler; `NewCSV adds EndDeviceProfileFetcher to the CSV converter`.

nicholaspcr replied:
Tried to make it simpler but I wasn't sure if the word `adds` fits the description, tried to word in a different way.

KrishnaIyer commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -297,6 +314,28 @@ var csvFieldSetters = map[string]csvFieldSetterFunc{
+		if dst.VersionIds == nil {
+			dst.VersionIds = &ttnpb.EndDeviceVersionIdentifiers{}
+		}
+		s, err := strconv.ParseUint(val, 10, 32)
```
Is this field defined in our CSV reference as decimal?
<img width="745" alt="Screenshot 2022-08-03 at 14 10 47" src="https://user-images.githubusercontent.com/13186113/182604517-4001a09a-28d7-4721-82c1-ea436576cdd7.png">

In the TR005 spec, these values are hexadecimal.

Same for `vendor_profile_id ` below.

KrishnaIyer commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -402,6 +465,24 @@ func (*ttsCSV) Convert(ctx context.Context, r io.Reader, ch chan<- *ttnpb.EndDev
+}
+
+// FromContextProfileIds fetches an EndDeviceProfileIdentifiers from within the given context, nil if does not exist.
+func FromContextProfileIds(ctx context.Context) *ttnpb.EndDeviceProfileIdentifiers {
```
This name does not follow our convention. Use `ProfileIDsFromContext`

KrishnaIyer commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -402,6 +465,24 @@ func (*ttsCSV) Convert(ctx context.Context, r io.Reader, ch chan<- *ttnpb.EndDev
+var profileIdsKey = &profileIdentifiersKeyType{}
+
+// NewContextWithProfileIds returns a context with a EndDeviceProfileIdentifiers injected into it.
+func NewContextWithProfileIds(ctx context.Context, ids *ttnpb.EndDeviceProfileIdentifiers) context.Context {
```
`NewContextWithProfileIDs`

KrishnaIyer commented on pkg/devicetemplates/ttscsv_test.go:
```diff
@@ -235,3 +352,30 @@ func TestTTSCSVConverter(t *testing.T) {
+
+func (*mockProfileFetcher) ShouldFetchProfile(device *ttnpb.EndDevice) bool {
+	return device.GetVersionIds().GetVendorId() != 0 &&
+		device.GetVersionIds().GetVendorProfileId() != 0
```
VendorProfileID can be zero. Only `VendorID` should be non zero.

KrishnaIyer commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -366,10 +405,21 @@ func (*ttsCSV) Convert(ctx context.Context, r io.Reader, ch chan<- *ttnpb.EndDev
 		}
-		paths := make([]string, 0, len(record))
+		// if fallback is set paths should be updated
+		if t.profile != nil && t.profile.ShouldFetchProfile(dev) {
+			paths = ttnpb.AddFields(paths, "version_ids.vendor_id", "version_ids.vendor_profile_id")
+		}
```
You can remove this extra condition and append these paths before `L552` as
```
tmpl.FieldMask.Paths = ttnpb.AddFields(tmpl.FieldMask.Paths, "version_ids.vendor_id", "version_ids.vendor_profile_id")
```

KrishnaIyer commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -402,6 +465,24 @@ func (*ttsCSV) Convert(ctx context.Context, r io.Reader, ch chan<- *ttnpb.EndDev
 
+type profileIdentifiersKeyType struct{}
+
+var profileIdsKey = &profileIdentifiersKeyType{}
```
`profileIDs`; it's unavoidable with protobuf generated code, but in hand-written code, we can use `IDs`.

KrishnaIyer commented on pkg/devicetemplateconverter/profile_fetcher.go:
```diff
@@ -0,0 +1,56 @@
+// ShouldFetchProfile dictactes if the end device has the necessary fields to fetch its profile.
+func (*ProfileFetcher) ShouldFetchProfile(device *ttnpb.EndDevice) bool {
+	return device.GetVersionIds().GetVendorId() != 0 &&
+		device.GetVersionIds().GetVendorProfileId() != 0
```
VendorProfileID can be zero. Only VendorID should be non zero.

KrishnaIyer commented on pkg/devicetemplateconverter/profile_fetcher.go:
```diff
@@ -0,0 +1,72 @@
+
+// FetchProfile provides the end device profile.
+func (pf *ProfileFetcher) FetchProfile(ctx context.Context, device *ttnpb.EndDevice) (*ttnpb.EndDeviceTemplate, error) {
+	if pf.DeviceRepository != nil {
+		return pf.DeviceRepository.GetTemplate(ctx, &ttnpb.GetTemplateRequest{
+			EndDeviceProfileIds: &ttnpb.EndDeviceProfileIdentifiers{
+				VendorId:        device.GetVersionIds().GetVendorId(),
+				VendorProfileId: device.GetVersionIds().GetVendorProfileId(),
+			},
+		})
+	}
+
+	conn, err := pf.Component.GetPeerConn(ctx, ttnpb.ClusterRole_DEVICE_REPOSITORY, nil)
+	if err != nil {
+		log.FromContext(ctx).WithError(err).Warn("Failed to get Device Repository peer")
+		return nil, err
+	}
+
+	opt, err := rpcmetadata.WithForwardedAuth(ctx, pf.Component.AllowInsecureForCredentials())
+	if err != nil {
+		return nil, err
+	}
+
+	return ttnpb.NewDeviceRepositoryClient(conn).GetTemplate(ctx, &ttnpb.GetTemplateRequest{
+		EndDeviceProfileIds: &ttnpb.EndDeviceProfileIdentifiers{
+			VendorId:        device.GetVersionIds().GetVendorId(),
+			VendorProfileId: device.GetVersionIds().GetVendorProfileId(),
+		},
+	}, opt)
```
Duplication of `GetTemplate` call isn't necessary. 
You're basically switching on the client connection so only that needs to be conditional.

nicholaspcr replied:
Not so sure about this, if I were to do something like:
```go
client, err := func() (ttnpb.DeviceRepositoryClient, error) {
if pf.DeviceRepository != nil {
	return pf.DeviceRepository, nil
}

conn, err := pf.Component.GetPeerConn(ctx, ttnpb.ClusterRole_DEVICE_REPOSITORY, nil)
if err != nil {
	log.FromContext(ctx).WithError(err).Warn("Failed to get Device Repository peer")
	return nil, err
}

return ttnpb.NewDeviceRepositoryClient(conn), nil
}()
if err != nil {
return nil, err
}

opt, err := rpcmetadata.WithForwardedAuth(ctx, pf.Component.AllowInsecureForCredentials())
if err != nil {
return nil, err
}
return client.GetTemplate(ctx, &ttnpb.GetTemplateRequest{
EndDeviceProfileIds: &ttnpb.EndDeviceProfileIdentifiers{
	VendorId:        device.GetVersionIds().GetVendorId(),
	VendorProfileId: device.GetVersionIds().GetVendorProfileId(),
},
}, opt)
```
There is a situation in which we are injecting the `WithForwardedAuth` into my request, which isn't a big problem but a bit I'm not sure if I should be forcing that into the test.

KrishnaIyer commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -393,6 +438,20 @@ func (*ttsCSV) Convert(ctx context.Context, r io.Reader, ch chan<- *ttnpb.EndDev
+			if err := profileTmpl.EndDevice.SetFields(tmpl.EndDevice, tmpl.FieldMask.Paths...); err != nil {
+				return err
+			}
+			tmpl.FieldMask.Paths = ttnpb.AddFields(tmpl.FieldMask.Paths, "version_ids.vendor_id", "version_ids.vendor_profile_id")
```
Shouldn't you copy the `version_ids` also to `profileTmpl`? This line should be before the `SetFields` right?

nicholaspcr replied:
You are correct, kind of an embarrassing mistake

KrishnaIyer reviewed (changes requested):
Looks much better!

nicholaspcr committed:
cli: Add version identifiers flagset to from-data command

nicholaspcr commented:
Recent changes on this PR.

@adriansmares pointed out something pertinent, the end-device payload formatter should be set to point to the DR. While there is some handling on the [backend](https://github.com/TheThingsNetwork/lorawan-stack/blob/ead0a3b40dc6d9bd3f2a524a72d473a34c4b6ccb/pkg/devicerepository/store/profiles.go#L103) that automatically does this, it only happens when if the `info` parameter is set and from the references bellow I couldn't find it being set on the remote or local stores.

- [DR-grpc](https://github.com/TheThingsNetwork/lorawan-stack/blob/ead0a3b40dc6d9bd3f2a524a72d473a34c4b6ccb/pkg/devicerepository/grpc.go#L183)
- [remote-store](https://github.com/TheThingsNetwork/lorawan-stack/blob/ead0a3b40dc6d9bd3f2a524a72d473a34c4b6ccb/pkg/devicerepository/store/remote/remote.go#L260-L264)
- [local-store](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.21/pkg/devicerepository/store/bleve/store.go#L187)

I ended up taking the simpler approach of just adding the formatter to the fetched profile.

CC: @adriansmares @KrishnaIyer

adriansmares commented on cmd/ttn-lw-cli/commands/end_device_templates.go:
```diff
@@ -309,10 +322,15 @@ This command takes end device templates from stdin.`,
+				EndDeviceProfileIds: nil,
+			}
+			// The execution should not stop if the flags are not found, therefore the error is ignored.
+			_, _ = req.EndDeviceProfileIds.SetFromFlags(cmd.Flags(), "version-ids")
```
Since the request field is `end-device-profile-ids`, let's keep that here too instead of `version-ids`.

adriansmares commented on cmd/ttn-lw-cli/commands/end_device_templates.go:
```diff
@@ -454,6 +475,7 @@ func init() {
 	endDeviceTemplatesCommand.AddCommand(endDeviceTemplatesListFormats)
 	endDeviceTemplatesFromDataCommand.Flags().AddFlagSet(templateFormatIDFlags())
 	endDeviceTemplatesFromDataCommand.Flags().AddFlagSet(dataFlags("", ""))
+	ttnpb.AddSetFlagsForEndDeviceProfileIdentifiers(endDeviceTemplatesFromDataCommand.Flags(), "version-ids", false)
```
Same here.

adriansmares commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -393,6 +438,33 @@ func (*ttsCSV) Convert(ctx context.Context, r io.Reader, ch chan<- *ttnpb.EndDev
+			if err != nil {
+				return err
+			}
+			// Set formatters to be form Device Repository.
```
Drop this for now.

nicholaspcr replied:
Dropped, the following FieldMasks remain on this section:
```
"version_ids.vendor_id",
"version_ids.vendor_profile_id",
```
They are accounted for not when importing the end-device profile but when the vendor_id is provided through the request body ( a fallback is provided ).


Regarding the discussion of making the DR set the formatters instead of leaving it be handled on the converter. I'll look into how to make the change and I'll write an issue about it. Then we can go back and forward there if the approach is adequate. What you think? @adriansmares

adriansmares replied:
> Regarding the discussion of making the DR set the formatters instead of leaving it be handled on the converter. I'll look into how to make the change and I'll write an issue about it. Then we can go back and forward there if the approach is adequate. What you think? @adriansmares

This is fine with me.

nicholaspcr commented:
Blocking this since there are some chances that were discussed with @adriansmares that are necessary for this PR, but for now I'm prioritizing the entity limit notification.

The changes discussed were:
- Removing the fieldMask added on this PR (should no longer be needed)
- Update the Device Repository in order to get the full versionID when making the request only with `vendorID` and `vendorProfileID`.
- Return the DR as a formatter when applicable ( this when making the request  only with `vendorID` and `vendorProfileID` )
- Update the API to use `oneof` in order to also accept passing the full `versionID`, instead of just accepting the `endDeviceProfileID`

johanstokking commented:
> * Update the Device Repository in order to get the full versionID when making the request only with `vendorID` and `vendorProfileID`.

Unfortunately this isn't possible.

The vendor ID can be mapped 1:0..1 with brands, that's the easy part.

However, many devices can use the same LoRaWAN device profile and codec. This is a feature of the Device Repository: declare the profile and codec once, and reference many times.

We cannot turn this around. We cannot go from a device profile (by ID) to versions. That would be a 1:N relation: we could return all version IDs that reference the profile, but we cannot return the exact version IDs.

So, on import, we can't use vendor ID and profile ID as keys. We may only go as far as getting the device profile by vendor ID and profile ID (so LoRaWAN version, RP version, class B/C parameters etc), which is still helpful.

nicholaspcr committed:
dev,util: Add end device profile fetcher to csv converter pkg

nicholaspcr commented:
> > * Update the Device Repository in order to get the full versionID when making the request only with `vendorID` and `vendorProfileID`.
> 
> Unfortunately this isn't possible.
> 
> The vendor ID can be mapped 1:0..1 with brands, that's the easy part.
> 
> However, many devices can use the same LoRaWAN device profile and codec. This is a feature of the Device Repository: declare the profile and codec once, and reference many times.
> 
> We cannot turn this around. We cannot go from a device profile (by ID) to versions. That would be a 1:N relation: we could return all version IDs that reference the profile, but we cannot return the exact version IDs.
> 
> So, on import, we can't use vendor ID and profile ID as keys. We may only go as far as getting the device profile by vendor ID and profile ID (so LoRaWAN version, RP version, class B/C parameters etc), which is still helpful.

With this in mind I just changed the API so that the convert request receives the versionIDs and changed the validation.

In regards to this:
>We may only go as far as getting the device profile by vendor ID and profile ID (so LoRaWAN version, RP version, class B/C parameters etc), which is still helpful.

The same information should be able to be fetched when passing the other fields in the versionID, no? I ended up allowing the user to provide the `vendorID` and `profileID` but the way I see it the current profile fetching doesn't really use those two specific values since I believe the same data could be fetched while using the other fields.

johanstokking commented:
> The same information should be able to be fetched when passing the other fields in the versionID, no? I ended up allowing the user to provide the `vendorID` and `profileID` but the way I see it the current profile fetching doesn't really use those two specific values since I believe the same data could be fetched while using the other fields.

Yes right. But sometimes the version IDs aren't available. For instance, in the standard LoRa Alliance QR code, there's only vendor ID and profile ID, and all we can do is prefill the device profile. The user still has to select which device it is.

nicholaspcr committed:
dtc: Add profile fetcher by vendorIDs

nicholaspcr committed:
dtc: Add profile fetcher by versionIDs

nicholaspcr committed:
dtc: Update profile fetcher usage

nicholaspcr committed:
dtc: Update grpc

nicholaspcr commented:
> > The same information should be able to be fetched when passing the other fields in the versionID, no? I ended up allowing the user to provide the `vendorID` and `profileID` but the way I see it the current profile fetching doesn't really use those two specific values since I believe the same data could be fetched while using the other fields.
> 
> Yes right. But sometimes the version IDs aren't available. For instance, in the standard LoRa Alliance QR code, there's only vendor ID and profile ID, and all we can do is prefill the device profile. The user still has to select which device it is.

With this in mind I opted to change things a bit, hopefully its more appropriate this way. I wrote a second fetcher on DTC that only takes into consideration the vendorID/ProfileID and instead of providing one fetcher to TTSCSV I pass both the vendorID fetcher and the versionID fetchers.

nicholaspcr commented on pkg/devicetemplateconverter/grpc.go:
```diff
@@ -39,25 +43,30 @@ func (s *endDeviceTemplateConverterServer) ListFormats(ctx context.Context, _ *p
 		case tmpl, ok := <-ch:
 			if !ok {
-				return nil
+				return ctx.Err()
```
An extra observation for reviewers is that I changed this due to some testing locally where when converting one device what happened was that the channel was closed and there could be the scenario where the `ctx.Error` is set but the `case tmpl, ok := <-ch:` was chosen first and it being closed returned a non error when its supposed to do the opposite. 

`Select` might chose randomly if both cases happen at once. Considering that the `ctx.Err()` is `nil` if not set I think this is an appropriate solution.

adriansmares commented on pkg/devicetemplateconverter/vendor_profile_fetcher.go:
```diff
@@ -0,0 +1,59 @@
+
+// ShouldFetchProfile dictactes if the end device has the necessary fields to fetch its profile.
+func (*VendorIDProfileFetcher) ShouldFetchProfile(device *ttnpb.EndDevice) bool {
+	return device.GetVersionIds().GetVendorId() != 0
```
`VendorProfileId` zero is not valid, so that should be checked as well, right ?

nicholaspcr replied:
Went based on this: https://github.com/TheThingsNetwork/lorawan-stack/pull/5610#discussion_r936595166

Will check when updating the other comments if its valid or not.

johanstokking replied:
It's not a valid value indeed. Only non-zero should be fetched.

adriansmares commented on pkg/devicetemplateconverter/version_profile_fetcher.go:
```diff
@@ -0,0 +1,74 @@
+func (*VersionIDProfileFetcher) ShouldFetchProfile(device *ttnpb.EndDevice) bool {
+	return device.GetVersionIds().GetBrandId() != "" &&
+		device.GetVersionIds().GetModelId() != "" &&
+		device.GetVersionIds().GetHardwareVersion() != "" &&
```
The hardware version is actually not required by the device repository. For example Dragino LHT65 does not have one. 
As such, you may skip it from the check here I believe.

adriansmares commented on pkg/devicetemplates/ttscsv.go:
```diff
@@ -383,10 +421,14 @@ func (*ttsCSV) Convert(ctx context.Context, r io.Reader, ch chan<- *ttnpb.EndDev
 		}
 
+		paths := make([]string, 0, len(record))
+		// fallback value for end device profile fetching
```
Nit: uppercase start and dot at the end.

adriansmares commented on pkg/devicetemplates/ttscsv_test.go:
```diff
@@ -231,11 +363,40 @@ func TestTTSCSVConverter(t *testing.T) {
+func (*mockProfileFetcher) ShouldFetchProfile(device *ttnpb.EndDevice) bool {
+	return device.GetVersionIds().GetBrandId() != "" &&
+		device.GetVersionIds().GetModelId() != "" &&
+		device.GetVersionIds().GetHardwareVersion() != "" &&
```
You probably want to drop the hardware version check here too.

johanstokking commented on pkg/devicerepository/mock/mock.go:
```diff
@@ -0,0 +1,114 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mockdr contains the mock of a Device Repository Server.
+package mockdr
```
Should this be `mock`?

nicholaspcr replied:
I went with the same standard as the `mockis` found [here](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.21/pkg/identityserver/mock/mock.go)

Could be just `mock` but I think its somewhat easier to have the abbreviation alongside the `mock` keyword.

johanstokking commented on pkg/devicetemplateconverter/vendor_profile_fetcher.go:
```diff
@@ -0,0 +1,59 @@
+
+// VendorIDProfileFetcher handles the validation and fetching of end device profile in the Device Repository.
+type VendorIDProfileFetcher struct {
+	Component        Component
+	DeviceRepository ttnpb.DeviceRepositoryClient
```
There's quite some (duplicate) code below for trying to get the template from `DeviceRepository` field here and otherwise dial DR and get it from there. This seems only be necessary for testing. Also `Component` is quite a dependency here.

What you really need is an interface to get a template. You could have two implementations: one that gets a peer conn (and takes an interface to get the peer conn and whether insecure is allowed for creds), and another (only in testing) to get it from some mock.

nicholaspcr replied:
I agree, with this in mind I took the approach of changing this quite a bit again. Made it to be its own package inside the DTC. Wrote the interface for just the `GetTemplate method` and basically removed the duplicate code from the fetchers' implementations

nicholaspcr committed:
dtc: Move profile fetcher implementation to its own package

nicholaspcr committed:
api: Add versionIDs to end-device template request

johanstokking commented:
@nicholaspcr please request review when this is ready for another pass.

johanstokking commented on cmd/ttn-lw-cli/commands/end_device_templates.go:
```diff
@@ -310,10 +323,15 @@ This command takes end device templates from stdin.`,
+				EndDeviceVersionIds: &ttnpb.EndDeviceVersionIdentifiers{},
+			}
+			// The execution should not stop if the flags are not found, therefore the error is ignored.
+			_, _ = req.EndDeviceVersionIds.SetFromFlags(cmd.Flags(), "end-device-version-ids")
```
nit; there's a difference between the flags not set (not changed) and failing to parse flags (i.e. wrong EUI length)

nicholaspcr replied:
True, changed it so that it there is an error it will return it. Indeed the flags should always be there for this command in specific since it is something declared at:
```
endDeviceTemplatesFromDataCommand.Flags().AddFlagSet(templateFormatIDFlags())
	endDeviceTemplatesFromDataCommand.Flags().AddFlagSet(dataFlags("", ""))
	ttnpb.AddSetFlagsForEndDeviceVersionIdentifiers(
		endDeviceTemplatesFromDataCommand.Flags(),
		"end-device-version-ids",
		false,
	)
```

johanstokking commented on pkg/devicetemplateconverter/devicetemplateconverter.go:
```diff
@@ -45,12 +46,24 @@ var errNotFound = errors.DefineNotFound("converter", "converter `{id}` not found
 
 // New returns a new *DeviceTemplateConverter.
 func New(c *component.Component, conf *Config) (*DeviceTemplateConverter, error) {
+	// Overwrites the TTSCSV converter in order to include a profile fetcher.
+	tmplFetcher, err := profilefetcher.NewTemplateFetcher(c)
+	if err != nil {
+		return nil, err
+	}
+	devicetemplates.RegisterConverter(
+		devicetemplates.TTSCSV,
+		devicetemplates.NewCSV(
+			profilefetcher.NewFetcherByVendorIDs(tmplFetcher),
+			profilefetcher.NewFetcherByVersionIDs(tmplFetcher),
+		),
+	)
```
What happens now is that a perceived singleton (the TTSCSV converter) is replaced by one that uses the last returned `profilefetcher.NewTemplateFetcher(c)`.

Instead, can we please put the fetcher in the context so that the singleton can extract it?

nicholaspcr replied:
A good idea, I took this a step further and changed so that there is no longer the `NewCSV` function and the CSV converter is declared how it previously was but now validates for the fetcher and if its `nil` the procedure is skipped. 
Updated the tests to accommodate for the changes as well, let me know what you think.

johanstokking reviewed (commented):
Looks great, few small things

nicholaspcr committed:
dtc, dev: Revert ttscsv to have private initialization

johanstokking commented on pkg/devicetemplateconverter/grpc.go:
```diff
@@ -39,25 +44,31 @@ func (s *endDeviceTemplateConverterServer) ListFormats(ctx context.Context, _ *p
 	ch := make(chan *ttnpb.EndDeviceTemplate)
 	go func() {
 		if err := converter.Convert(ctx, bytes.NewReader(req.Data), ch); err != nil {
 			cancel(err)
 		}
 	}()
+
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
 		case tmpl, ok := <-ch:
 			if !ok {
-				return nil
+				return ctx.Err()
```
Why is this?

This would be a race between `ch` close and `ctx` done, so it's probably negligble in practice. However, theoterically, if the producer is done and closes `ch`, I don't know why a `ctx` error would be relevant. Why would we timeout if we have everything?

nicholaspcr replied:
https://github.com/TheThingsNetwork/lorawan-stack/pull/5610#discussion_r949211645

The idea behind it was that even when there is an error the channel is closed and therefore it isn't guaranteed that the flow will be directed to:
```go
case <-ctx.Done():
	return ctx.Err()
```
So there could be cases where there is an error, the channel is closed by the converter when returning said error and the select would be somewhat ambiguous. Meaning that `ctx.Err` is defined but since the channel is closed it could enter the case statement below and not provide such error to the user.
```go
case tmpl, ok := <-ch:
	if !ok {
		return nil
	}
```

At least this was the explanation that made sense for the error that I encountered while testing. Considering that the `ctx.Error` is `nil` whenever it is not set I approached this in a way that I wouldn't change the code too much.

johanstokking commented on pkg/devicetemplates/ttscsv_test.go:
```diff
@@ -43,27 +47,30 @@ var (
+		name            string
+		reader          io.Reader
+		converter       Converter
+		contextInjector func(context.Context) context.Context
```
nit;
```suggestion
		fillContext func(context.Context) context.Context
```

KrishnaIyer reviewed (approved):
LGTM now
